src/subscription/__tests__/subscribe-test.js

```diff
@@ -617,33 +618,196 @@ describe('Subscribe', () => {
+
+    expect(result).to.deep.equal({
+      done: false,
+      value: {
+        errors: [
+          {
+            message: 'test error',
+            locations: [ { line: 3, column: 11 } ],
+            path: [ 'importantEmail' ]
+          }
+        ]
+      }
+    });
+
+    expect(
+      await subscription.next()
+    ).to.deep.equal({ value: undefined, done: true });
+  });
+
+  it('should report error returned by subscribe function', async () => {
+    const erroringEmailSchema = emailSchemaWithSubscribeFn(
+      function importantEmail() {
+        return new Error('test error');
+      }
+    );
+
+    const subscription = subscribe(
+      erroringEmailSchema,
+      parse(`
+        subscription {
+          importantEmail
+        }
+      `)
+    );
+
+    const result = await subscription.next();
+
+    expect(result).to.deep.equal({
+      done: false,
+      value: {
+        errors: [
+          {
+            message: 'test error',
+            locations: [ { line: 3, column: 11 } ],
+            path: [ 'importantEmail' ]
+          }
+        ]
+      }
+    });
+
+    expect(
+      await subscription.next()
+    ).to.deep.equal({ value: undefined, done: true });
+  });
+
+  it('should handle error during execuction of source event', async () => {
+    const erroringEmailSchema = new GraphQLSchema({
       query: QueryType,
       subscription: new GraphQLObjectType({
         name: 'Subscription',
         fields: {
           importantEmail: {
             type: GraphQLString,
-            subscribe: () => {
-              throw new Error('test error');
+            resolve(event) {
+              if (event === 'Goodbye') {
+                throw new Error('Never leave.');
+              }
+              return event;
+            },
+            subscribe: async function* importantEmail() {
+              yield 'Hello';
+              yield 'Goodbye';
             },
           },
         },
       })
     });
 
-    const ast = parse(`
-      subscription {
-        importantEmail
+    const subscription = subscribe(
+      erroringEmailSchema,
+      parse(`
+        subscription {
+          importantEmail
+        }
+      `)
+    );
+
+    const payload1 = await subscription.next();
+    expect(payload1).to.jsonEqual({
+      done: false,
+      value: {
+        data: {
+          importantEmail: 'Hello'
+        }
       }
-    `);
+    });
 
-    expect(() => {
-      subscribe(
-        invalidEmailSchema,
-        ast
-      );
-    }).to.throw('test error');
+    const payload2 = await subscription.next();
+    expect(payload2).to.jsonEqual({
+      done: false,
+      value: {
+        errors: [
+          {
+            message: 'Never leave.',
+            locations: [ { line: 3, column: 11 } ],
+            path: [ 'importantEmail' ],
+          }
+        ],
+        data: {
+          importantEmail: null,
+        }
+      }
+    });
+  });
+
+  function emailSchemaWithSubscribeFn(subscribeFn) {
```

Comment:
Nit: this is a weird place for this function. Perhaps move it to the top of the file or before the block of tests that use it?